### PR TITLE
Add support for multiple URL's in entry

### LINKED ISSUE #398: Support multiple urls per entry
Hey,

is it possible to add multiple urls to a single entry?
Maybe a cool feature?
(Working with passiFox)

Greetings Kaito

### PR DESCRIPTION
**Pull request by varjolintu**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
Adds support for multiple URL's per entry. The new Browser Integration entry settings page has a list view with any additional URL's. These URL's are added to the entry attributes with `KP2A_URL_<counter>`, which means those are directly compatible with Keepass2Android.

Please note that #3444 (Browser Integration releated entry settings) must be merged before this PR. This extends that feature. Alternatively, that PR can be dismissed totally and all the features will be included in this one.

Fixes #398.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

The Browser Integration page shows the following list:
Screen Shot 2019-09-19 at 12 39 41

The URL's will be also added to the attributes page:
Screen Shot 2019-09-19 at 12 39 52

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Manually. GUI tests are also included.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/browser/BrowserService.cpp [hunk 1]
```diff
@@ -50,6 +50,9 @@ static int KEEPASSXCBROWSER_DEFAULT_ICON = 1;
 const char BrowserService::LEGACY_ASSOCIATE_KEY_PREFIX[] = "Public Key: ";
 static const char KEEPASSHTTP_NAME[] = "KeePassHttp Settings";
 static const char KEEPASSHTTP_GROUP_NAME[] = "KeePassHttp Passwords";
+// Extra entry related options saved in custom data
+const char BrowserService::OPTION_SKIP_AUTO_SUBMIT[] = "BrowserSkipAutoSubmit";
```
Comment (droidmonkey) [inline src/browser/BrowserService.cpp]: For these you can also use a QStringLiteral which builds a QString object at compile time. This is my preferred method over cstrings

File: src/browser/BrowserService.cpp [hunk 2]
```diff
@@ -590,13 +597,42 @@ BrowserService::searchEntries(const QSharedPointer<Database>& db, const QString&
...
+    };
+
+    // Search host from URL fields
+    for (Entry* entry : EntrySearcher().search(baseDomain(hostname), rootGroup)) {
+        if (!handleURL(entry->url())) {
+            continue;
+        }
+
+        entries.append(entry);
+    }
```
Comment (droidmonkey) [inline src/browser/BrowserService.cpp]: const auto&

File: src/gui/entry/EditEntryWidget.cpp [hunk 3]
```diff
@@ -246,6 +261,121 @@ void EditEntryWidget::setupAutoType()
...
+    auto hide = m_browserUi->hideEntryCheckbox->isChecked();
+    m_customData->set(BrowserService::OPTION_SKIP_AUTO_SUBMIT, (skip ? QString("true") : QString("false")));
+    m_customData->set(BrowserService::OPTION_HIDE_ENTRY, (hide ? QString("true") : QString("false")));
+}
+
+void EditEntryWidget::insertURL()
+{
+    Q_ASSERT(!m_history);
+
+    QString name = tr("KP2A_URL");
```
Comment (droidmonkey) [inline src/gui/entry/EditEntryWidget.cpp]: We don't want to translate this

File: src/gui/entry/EditEntryWidget.cpp [hunk 4]
```diff
@@ -246,6 +261,121 @@ void EditEntryWidget::setupAutoType()
...
+{
+    m_browserSettingsChanged = true;
+}
+
+void EditEntryWidget::updateBrowser()
+{
+    if (!m_browserSettingsChanged) {
+        return;
+    }
+
```
Comment (droidmonkey) [inline src/gui/entry/EditEntryWidget.cpp]: QStringLiteral

File: src/gui/entry/EditEntryWidget.cpp [hunk 5]
```diff
@@ -330,6 +460,16 @@ void EditEntryWidget::setupEntryUpdate()
         connect(m_sshAgentUi->lifetimeSpinBox, SIGNAL(valueChanged(int)), this, SLOT(setModified()));
     }
```
Comment (droidmonkey) [inline src/gui/entry/EditEntryWidget.cpp]: You can remove the "this", it makes it a little cleaner to read

File: src/browser/BrowserService.cpp (hunk 1 above)
Comment (varjolintu) [reply src/browser/BrowserService.cpp]: Only reason that was not used is that QLatin1String doesn't handle QStrings directly. If we must prefer QStringLiteral, we must make an additional wrapper function for QLatin1String.

File: src/browser/BrowserService.cpp (hunk 2 above)
Comment (varjolintu) [reply src/browser/BrowserService.cpp]: It's QList<Entry*> everywhere. We really want to modify everything to QList<const Entry*>?

Comment (droidmonkey) [reply src/browser/BrowserService.cpp]: Standby I think my comment went to the wrong line

File: src/gui/entry/EditEntryWidget.cpp (hunk 4 above)
Comment (varjolintu) [reply src/gui/entry/EditEntryWidget.cpp]: What's the correct line here?

File: src/browser/BrowserService.cpp (hunk 1 above)
Comment (varjolintu) [reply src/browser/BrowserService.cpp]: Using `.toUtf8()` for `QStringLiteral` was enough.

**Comment by varjolintu** [comment]
@droidmonkey Fixes made.

File: src/browser/BrowserService.cpp [hunk 6]
```diff
@@ -317,7 +320,7 @@ QString BrowserService::storeKey(const QString& key)
             return {};
         }
 
-        contains = db->metadata()->customData()->contains(QLatin1String(ASSOCIATE_KEY_PREFIX) + id);
+        contains = db->metadata()->customData()->contains(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + id);
```
Comment (yan12125) [inline src/browser/BrowserService.cpp]: I think `QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8())` can simply be `ASSOCIATE_KEY_PREFIX`?

**Comment by yan12125** [comment]
An untested diff for my previous comment:
```Diff
diff --git a/src/browser/BrowserService.cpp b/src/browser/BrowserService.cpp
index eb4c32b3..edcd793e 100644
--- a/src/browser/BrowserService.cpp
+++ b/src/browser/BrowserService.cpp
@@ -320,7 +320,7 @@ QString BrowserService::storeKey(const QString& key)
             return {};
         }

-        contains = db->metadata()->customData()->contains(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + id);
+        contains = db->metadata()->customData()->contains(ASSOCIATE_KEY_PREFIX + id);
         if (contains) {
             dialogResult = MessageBox::warning(nullptr,
                                                tr("KeePassXC: Overwrite existing key?"),
@@ -333,7 +333,7 @@ QString BrowserService::storeKey(const QString& key)
     } while (contains && dialogResult == MessageBox::Cancel);

     hideWindow();
-    db->metadata()->customData()->set(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + id, key);
+    db->metadata()->customData()->set(ASSOCIATE_KEY_PREFIX + id, key);
     return id;
 }

@@ -344,7 +344,7 @@ QString BrowserService::getKey(const QString& id)
         return {};
     }

-    return db->metadata()->customData()->value(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + id);
+    return db->metadata()->customData()->value(ASSOCIATE_KEY_PREFIX + id);
 }

 QJsonArray BrowserService::findMatchingEntries(const QString& id,
@@ -651,7 +651,7 @@ QList<Entry*> BrowserService::searchEntries(const QString& url, const StringPair
                     // Check if database is connected with KeePassXC-Browser
                     for (const auto& keyPair : keyList) {
                         QString key =
-                            db->metadata()->customData()->value(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + keyPair.first);
+                            db->metadata()->customData()->value(ASSOCIATE_KEY_PREFIX + keyPair.first);
                         if (!key.isEmpty() && keyPair.second == key) {
                             databases << db;
                         }
@@ -741,12 +741,9 @@ void BrowserService::convertAttributesToCustomData(const QSharedPointer<Database
         return;
     }

-    const QString keePassBrowserGroupName = QLatin1String(KEEPASSXCBROWSER_GROUP_NAME.toUtf8());
-    const QString keePassHttpGroupName = QLatin1String(KEEPASSHTTP_GROUP_NAME.toUtf8());
-
     for (auto* g : rootGroup->groupsRecursive(true)) {
-        if (g->name() == keePassHttpGroupName) {
-            g->setName(keePassBrowserGroupName);
+        if (g->name() == KEEPASSHTTP_GROUP_NAME) {
+            g->setName(KEEPASSXCBROWSER_GROUP_NAME);
             break;
         }
     }
@@ -919,17 +916,15 @@ Group* BrowserService::getDefaultEntryGroup(const QSharedPointer<Database>& sele
         return nullptr;
     }

-    const QString groupName = QLatin1String(KEEPASSXCBROWSER_GROUP_NAME.toUtf8());
-
     for (auto* g : rootGroup->groupsRecursive(true)) {
-        if (g->name() == groupName && !g->isRecycled()) {
+        if (g->name() == KEEPASSXCBROWSER_GROUP_NAME && !g->isRecycled()) {
             return db->rootGroup()->findGroupByUuid(g->uuid());
         }
     }

     auto* group = new Group();
     group->setUuid(QUuid::createUuid());
-    group->setName(groupName);
+    group->setName(KEEPASSXCBROWSER_GROUP_NAME);
     group->setIcon(KEEPASSXCBROWSER_DEFAULT_ICON);
     group->setParent(rootGroup);
     return group;
@@ -1100,8 +1095,8 @@ int BrowserService::moveKeysToCustomData(Entry* entry, const QSharedPointer<Data
             publicKey.remove(LEGACY_ASSOCIATE_KEY_PREFIX);

             // Add key to database custom data
-            if (db && !db->metadata()->customData()->contains(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + publicKey)) {
-                db->metadata()->customData()->set(QLatin1String(ASSOCIATE_KEY_PREFIX.toUtf8()) + publicKey,
+            if (db && !db->metadata()->customData()->contains(ASSOCIATE_KEY_PREFIX + publicKey)) {
+                db->metadata()->customData()->set(ASSOCIATE_KEY_PREFIX + publicKey,
                                                   entry->attributes()->value(key));
                 ++keyCounter;
             }
```

By the way, could you rebase this PR against the `develop` branch?

**Comment by droidmonkey** [comment]
I rebased, this could use a bit of testing before merge. Please give me thumbs up when ready.

File: src/browser/BrowserService.cpp [hunk 7]
```diff
@@ -598,13 +597,42 @@ BrowserService::searchEntries(const QSharedPointer<Database>& db, const QString&
...
+                continue;
+            }
+
+            for (const auto& key : entry->attributes()->keys()) {
+                if (key.startsWith("KP2A_URL") && handleURL(entry->attributes()->value(key))) {
+                    entries.append(entry);
+                }
+            }
         }
     }
```
Comment (droidmonkey) [inline src/browser/BrowserService.cpp]: These loops can be combined. You are effectively recursing through the entire database twice. I wouldn't use the entry searcher in this case.

Comment (varjolintu) [reply src/browser/BrowserService.cpp]: I agree. Two loops are not necessary. The reasong this was left this way was that EntrySearcher didn't support such regex based search based on attribute's name. I'll clean this up.

And writing some extra non-GUI tests would be also a nice addition.

Comment (varjolintu) [reply src/browser/BrowserService.cpp]: Changes made.

**Comment by phoerious** [comment]
Could we integrate this with the main edit entry widget? I think it makes more sense to support multiple URLs in general than to support it only for browser integration.

Also, I don't really like where this is going with all the vendor prefixes for custom attributes. Should we be initiating a joint effort with other programs to agree on an extended set of attributes with a common prefix? Otherwise this will get very chaotic. There's a reason why CSS got rid of vendor prefixes.

**Comment by varjolintu** [comment]
@phoerious We could. If we are going to do that, do we maintain the current UI or how are we going to display/add/modify/delete the additional URL's in the main edit entry widget?

I agree that a common prefix with other programs would be a much better choice. AFAIK the only application supporting multiple URL's at this point is KP2A, so it makes sense (at least for me) to use that prefix for now, even if it's a custom one.

**Comment by yan12125** [comment]
> AFAIK the only application supporting multiple URL's at this point is KP2A

Another one that supports multiple URLs in an entry is Kee (formerly KeeFox). During the last time I checked (two years ago), it uses a quite different, JSON-based format as I described in https://github.com/keepassxreboot/keepassxc/issues/398#issuecomment-326755012.

**Comment by varjolintu** [comment]
@yan12125 Thanks. Good to know. It seems a bit complex, if the main function is just to provide an additional URL('s).

**Comment by phoerious** [comment]
No JSON, please. The KDBX format is very flexible in what attributes it allows and the only difference between "main" attributes and "custom" attributes is that the GUI knows how to interpret the "main" ones. We should make use of that.

**Comment by droidmonkey** [comment]
So are we good with this for now? We can always work with KeePass2Android to define a more common method. Ideally this would also be a starting point to a better entry template format as well.

**Comment by varjolintu** [comment]
Rebased.

**Comment by AlexanderAmelkin** [comment]
Great! Very good you made it compatible with KP2A! Unfortunately, editing the URLs via the 'Browser integration' tab doesn't work for me. I always end up with a `<Пустой URL>` (Empty URL in Russian) after editing it. I can successfully edit the `KP2A_URL_*` additional fields though and the edited content is then properly displayed in the "Browser integration" tab.

**Comment by opqriu** [comment]
While using Appimage in Debian I found an issue where Multiple URLs were not possible. So I came here. Do I need to compile and install keepassxreboot to use this version?

**Comment by varjolintu** [comment]
@opqriu What version are you using? This is introduced in 2.5.0.

**Comment by opqriu** [comment]
Thanks. I solved I found it. Good day.

**Comment by BaileyLove** [comment]
Confused. Installed the current (2.5.3) version of KeePassXC and the "Browser Integration" thing is just not there. I have it in settings, but not for entries. Was this feature removed? Is KeePassReboot not the same software as KeePassXC? What am I missing? Ty.

**Comment by stratozero** [comment]
Is it possible to use wildcard or regular expressions in additional URLs? Or should I add every additional URL as an entry?
e.g.: using _https://login.domain.com_ as URL, then _https://*.domain.com_ as custom URL to match every other, so that when I click on the entry i get redirected to the login page, but also other website with the same second level domain are matched by the browser pugin?

**Comment by varjolintu** [comment]
> Is it possible to use wildcard or regular expressions in additional URLs? Or should I add every additional URL as an entry? e.g.: using _https://login.domain.com_ as URL, then _https://*.domain.com_ as custom URL to match every other, so that when I click on the entry i get redirected to the login page, but also other website with the same second level domain are matched by the browser pugin?

https://github.com/keepassxreboot/keepassxc/pull/9835

**Pull request merged**